[quant][pt2e][refactor] Move the annotation for observer sharing ops into separate util

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99384

Summary:
In order to keep quantizer simple, we want to move the annotation code for operators like flatten, hardtanh etc. to
a separate utility function that is called after the quantizer annotation is done, this makes these ops (operator list) not
configurable by user, and also makes prepare_pt2e operator aware instead of operator agnostic, this design is not final,
we may change it in the future if we find there are use cases that need these to be configurable or if we feel it is important for prepare_pt2e
to stay agnostic to operator/operator patterns

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_obs_sharing_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45071006](https://our.internmc.facebook.com/intern/diff/D45071006)

jerryzh168 committed:
[quant][pt2e][refactor] Move the annotation for observer sharing ops into separate util

Summary:
In order to keep quantizer simple, we want to move the annotation code for operators like flatten, hardtanh etc. to
a separate utility function that is called after the quantizer annotation is done, this makes these ops (operator list) not
configurable by user, and also makes prepare_pt2e operator aware instead of operator agnostic, this design is not final,
we may change it in the future if we find there are use cases that need these to be configurable or if we feel it is important for prepare_pt2e
to stay agnostic to operator/operator patterns

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_obs_sharing_ops

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45071006).

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Move the annotation for observer sharing ops into separate util"


Summary:
In order to keep quantizer simple, we want to move the annotation code for operators like flatten, hardtanh etc. to
a separate utility function that is called after the quantizer annotation is done, this makes these ops (operator list) not
configurable by user, and also makes prepare_pt2e operator aware instead of operator agnostic, this design is not final,
we may change it in the future if we find there are use cases that need these to be configurable or if we feel it is important for prepare_pt2e
to stay agnostic to operator/operator patterns

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_obs_sharing_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45071006](https://our.internmc.facebook.com/intern/diff/D45071006)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45071006).

leslie-fang-intel commented on torch/ao/quantization/_pt2e/_propagate_annotation.py:
```diff
@@ -0,0 +1,45 @@
+            continue
+
+        # make sure current node is not annotated
+        if "target_dtype_info" in n.meta and n.meta["target_dtype_info"].get("_annotated", False):
```
From current implementation, looks like these ops in `_is_share_obs_or_fq_op` is still configurable by user. It will skip the `propagate_annotation` if it has already be annotated in `quantizer`. Should we add some check to prevent these ops accidently annotated by `quantizer`?

jerryzh168 replied:
I think that's allowed, it's fine if user wants to add some annotation for these ops, in that case we just don't do propagation. what I meant by not configurable is that the op_list is now fixed, user can't add new ops to this list right now

leslie-fang-intel replied:
Got it. Make sense and thanks.

kimishpatel reviewed (approved):
This looks good to me.

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Move the annotation for observer sharing ops into separate util"


Summary:
In order to keep quantizer simple, we want to move the annotation code for operators like flatten, hardtanh etc. to
a separate utility function that is called after the quantizer annotation is done, this makes these ops (operator list) not
configurable by user, and also makes prepare_pt2e operator aware instead of operator agnostic, this design is not final,
we may change it in the future if we find there are use cases that need these to be configurable or if we feel it is important for prepare_pt2e
to stay agnostic to operator/operator patterns

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_obs_sharing_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45071006](https://our.internmc.facebook.com/intern/diff/D45071006)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Move the annotation for observer sharing ops into separate util"


Summary:
In order to keep quantizer simple, we want to move the annotation code for operators like flatten, hardtanh etc. to
a separate utility function that is called after the quantizer annotation is done, this makes these ops (operator list) not
configurable by user, and also makes prepare_pt2e operator aware instead of operator agnostic, this design is not final,
we may change it in the future if we find there are use cases that need these to be configurable or if we feel it is important for prepare_pt2e
to stay agnostic to operator/operator patterns

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_obs_sharing_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45071006](https://our.internmc.facebook.com/intern/diff/D45071006)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Move the annotation for observer sharing ops into separate util"


Summary:
In order to keep quantizer simple, we want to move the annotation code for operators like flatten, hardtanh etc. to
a separate utility function that is called after the quantizer annotation is done, this makes these ops (operator list) not
configurable by user, and also makes prepare_pt2e operator aware instead of operator agnostic, this design is not final,
we may change it in the future if we find there are use cases that need these to be configurable or if we feel it is important for prepare_pt2e
to stay agnostic to operator/operator patterns

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_obs_sharing_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45071006](https://our.internmc.facebook.com/intern/diff/D45071006)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Move the annotation for observer sharing ops into separate util"


Summary:
In order to keep quantizer simple, we want to move the annotation code for operators like flatten, hardtanh etc. to
a separate utility function that is called after the quantizer annotation is done, this makes these ops (operator list) not
configurable by user, and also makes prepare_pt2e operator aware instead of operator agnostic, this design is not final,
we may change it in the future if we find there are use cases that need these to be configurable or if we feel it is important for prepare_pt2e
to stay agnostic to operator/operator patterns

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_obs_sharing_ops

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45071006](https://our.internmc.facebook.com/intern/diff/D45071006)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45071006).
